opentelemetry-api/src/opentelemetry/trace/propagation/textmap.py

```diff
@@ -20,7 +20,64 @@
 TextMapPropagatorT = typing.TypeVar("TextMapPropagatorT")
 
 Setter = typing.Callable[[TextMapPropagatorT, str, str], None]
-Getter = typing.Callable[[TextMapPropagatorT, str], typing.List[str]]
+
+
+def default_get(carrier: TextMapPropagatorT, key: str) -> typing.List[str]:
```

Comment:
Why do we have default implementations for `get` and `keys`? Where is this implementation from?

Reply:
this is implemented as @owais suggested, https://github.com/open-telemetry/opentelemetry-python/pull/1196#discussion_r501678110


and implementation of default function is taken as this was the implementation in most of the files.


I'm facing a typing issue with TextMapPropagatorT type, Could you please help with this, details can be found in this comment 
https://github.com/open-telemetry/opentelemetry-python/pull/1196#issuecomment-706678858

Reply:
Took a quick look. Your implementation is accepting a TextMapPropagatorT type but then using it like a dictionary resulting in the error. Do we need a default implementation here? Suggest removing it and keeping only interfaces here. The SDK package may contain a carrier specific implementation (one for dictionaries for example) if a common one is used across the project.

Reply:
@owais If we remove the default implementation, all the files using getter will have to implement keys even if they don't use it.

Reply:
To fix your typing errors, they should be implementation specific. AFAIU, what we need is a dict implementation of getter as that is used almost everywhere. Such an implementation should be specific to dictionaries, i.e, accept dictionaries as the carrier argument, not the generic var.

Reply:
@owais 
Should we keep the default implementations then? (common across many instrumentations)

Reply:
Yes, we can keep it. My comment was more about keeping it in API vs moving to SDK. I'm fine with either.

Reply:
> Took a quick look. Your implementation is accepting a TextMapPropagatorT type but then using it like a dictionary resulting in the error. Do we need a default implementation here? Suggest removing it and keeping only interfaces here. The SDK package may contain a carrier specific implementation (one for dictionaries for example) if a common one is used across the project.

@owais I'm a bit confused with this one, here is my understanding

- If we are going to remove the default implementation we have to implement get() and keys() everywhere. 
If we go with keeping the default implementation in textmap.py we will have a typing error because the default implementation uses a dict type and carrier expects a type of generic var(TextMapPropagatorT). If we have to change the type of the get method of getter to  `typing.Callable[[typing.Dict[str,str], str], typing.List[str]]` we have to change the carrier type to Dict as well from generic var in all the implementations. 

what do you suggest we do?

Reply:
May be it is stemming from my little understanding of mypy but I think there are two issues overlapping here. 

1. Whether and where to keep a commonly used implementation for dicts as dicts seem to be most common carriers. Perhaps we shouldn't call it default implementation, just a dicts one. 
2. My original comment was about reducing boilerplate code in instrumentations and user code. The suggestion was to have a function or utility class that could turn simple lambdas or functions into fully compatible Getter implementations so that each instrumentation did not have to define a class and all methods to implement a getter. If  dict is being used in 99% of the cases and we just need to add custom getter implementations for in just one or two places then feel free to not add this.

I have a working example below. It defines the Getter interface and then implements two Getter implementations for List and Dict types. It also implements a "Helper" implementation (which I had originally requested) that can be used to turn any set of get and keys functions into a compatible getter. 

```python
from typing import TypeVar, Generic, List, Dict, Callable


CarrierT = TypeVar('CarrierT')


class Getter(Generic[CarrierT]):
    def get(self, carrier: CarrierT, key: str) -> List[str]:
        raise NotImplementedError()
    
    def keys(self, carrier: CarrierT) -> List[str]:
        raise NotImplementedError()
    

def print_carrier_contents(getter: Getter, carrier: CarrierT):
    for key in getter.keys(carrier):
        print(key, ' = ', getter.get(carrier, key))
    print()


try:
   print_carrier_contents(Getter(), {})
except NotImplementedError:
   pass

class ListGetter(Getter):
    def get(self, carrier: List[str], key: str) -> List[str]:
        for pair in carrier:
            if pair[0] == key:
                return [pair[1]]
        return []
    
    def keys(self, carrier: List[str]) -> List[str]:
        return [i[0] for i in carrier]
    
print_carrier_contents(ListGetter(), [("k1", "v1"), ("k2", "v2")])


class DictGetter(Getter):
    def get(self, carrier: Dict[str, str], key: str) -> List[str]:
        val = carrier.get(key, None)
        if val:
          return [val]
        return []
    
    def keys(self, carrier: Dict[str, str]) -> List[str]:
        return list(carrier.keys())

print_carrier_contents(DictGetter(), {"k1": "v1", "k2": "v2"})



class HelperGetter(Getter):

    def __init__(self, get: Callable[[CarrierT, str], List[str]], keys: Callable[[CarrierT], List[str]]):
      self._get = get
      self._keys = keys

    def get(self, carrier: CarrierT, key: str) -> List[str]:
      return self._get(carrier, key)

    def keys(self, carrier: CarrierT) -> List[str]:
      return self._keys(carrier)


class CustomCarrier:
    
    def get_value(self, key):
        if key == "k1":
            return ["v1"]
        elif key == "k2":
            return ["v2"]
        return []
        
    def list_keys(self):
        return ["k1", "k2"]
        
        
getter = HelperGetter(get=CustomCarrier.get_value, keys=CustomCarrier.list_keys)

print_carrier_contents(getter, CustomCarrier())
```

Hope this helps.